server/app.js

```diff
@@ -26,9 +25,46 @@ io.on('connection', socket => {
   socket.emit('updateUserList', users);
 
   socket.on('newUser', user => {
-    const newUser = { username: user.username, color: user.color, id: socket.id };
+    const newUser = { username: user.username, color: user.color, id: socket.id, isReady: false };
     users.push(newUser);
     io.emit('updateUserList', users);
+    socket.emit('newUser', { userId: socket.id, playersReady });
+  });
+
+  // from the client ready / waiting status change the player status
+  socket.on('userStatus', status => {
+    // does user exist in users, if not exit function
+    const user = users.find(user => user.id === socket.id);
+    if (!user) return;
+
+    if (status.statusText === 'ready') {
+      if (!user.isReady) {
+        user.isReady = true;
+        playersReady += 1;
+      }
+    } else {
+      if (user.isReady) {
+        user.isReady = false;
+        playersReady -= 1;
+      }
+    }
+    io.emit('userStatus', status);
+    io.emit('playersReady', playersReady);
+  });
+
+  // on disconnect kick player from game
+  socket.on('disconnect', () => {
+    // find the userindex in the users array that corresponds to the disconnected socket.id
+    const userIndex = users.findIndex(user => user.id === socket.id);
+    if (userIndex === -1) return;
+    // if player was in ready move, remove their ready from the playersReady number
+    if (users[userIndex].isReady) {
+      playersReady -= 1;
+      io.emit('playersReady', playersReady);
+    }
+    // remove the user with splice
+    users.splice(userIndex, 1);
+    io.emit('updateUserList', users);
   });
 });
 
```

Comment:
Tänker att vi kanske behöver mer felhantering sen (inget som behöver ändras nu)
Funderar också på om vi borde refaktorera koden framöver och göra funktioner så att koden blir mer läsbar och tydlig (och återanvändbar), men kan inte tillräckligt mycket för att veta om det är bäst practice, det känns bara som att det skulle bli mer förståeligt (för mig). 🐙 

Min spontanta tanke vore att bryta ut funktioner:

t.ex. diverse funktioner:

function addUser() {...}
function removeUserOnDisconnect() {...}

Som sen används här i: 

io.on('connection', socket => {
 
  socket.on('newUser', user => {
    addUser(user);
    socket.emit('newUser', { userId: socket.id, playersReady ...
  });


  socket.on('disconnect', () => {
    removeUserOnDisconnect(socket.id);
  });

});

osv.

Reply:
Yes man skulle definitivt kunna bryta ner dessa i mindre beståndsdelar, är alltid svårt att väga in hur små dock, men ser man någon annans kod man själv skulle vilja bryta ner lite mer för att få förståelse så kan man göra detta tycker jag, är ju meningen att den ska vara läsbar, kan även gynna sedan i testning. 👍 😸.